### ISSUE TITLE
Should HealthCheckRoutes be enabled by default?

### ISSUE THREAD

**Issue by jroper** [akka/akka-management#470]
Currently they aren't, though the documentation implies they are, and makes no mention of having to add them to the route providers:

https://developer.lightbend.com/docs/akka-management/current/bootstrap/recipes.html

Also, the documentation for cluster health checks references `akka-cluster-http-management` in the wording, but then in the code sample, instructs to add the `akka-management` dependency. Actually, both of these are wrong, it should be `akka-management-cluster-http`.

**Comment by patriknw** [comment]
The reasons for not enabling it by default is that then it's not possible (or at least difficult) to disable or replace it. Therefore this must be configured:
```
akka.management.http.route-providers += "akka.management.HealthCheckRoutes"
```

That enables empty readiness and liveness checks, and allows to plug in custom checks.

Thereafter the Cluster membership readiness check can be added with dependency `akka-management-cluster-http`. It's by default adding `cluster-membership` check: https://github.com/akka/akka-management/blob/master/cluster-http/src/main/resources/reference.conf#L29
(this can be disabled by setting the property to `""` or replace with another implementation)

Looks like the documentation must be updated.

**Comment by chbatey** [comment]
Docs for health checks describe how to host them as an akka management route: https://developer.lightbend.com/docs/akka-management/current/healthchecks.html

**Comment by jrudolph** [comment]
Could we instead add the route by default and add another flag that guards whether the health-check is really enabled or not?

**Comment by ignasi35** [comment]
related https://github.com/lagom/lagom/pull/1738 

tl;dr; after a couple of iterations we're undecided on whether Lagom should enable the `HealthCheckRoutes` either (inclined to _not_ enable them on the framework but add the setting on the `g8` seeds [like other defaults](https://github.com/lagom/lagom-scala.g8/blob/master/src/main/g8/%24name__norm%24-impl/src/main/resources/application.conf#L31)).

**Comment by renatocaval** [comment]
I was thinking about that because we were in doubt if we should enable it by default in Lagom. 

I thought about two options: 

## Option 1: 
Add the means to remove a route like we have for Play modules:
   `akka.management.http.disabled.route-providers += "akka.management.HealthCheckRoutes"`

A little bit awkward because of lack of simmetry. 

## Option 2: 
Filter out empty values. That would allow us to do the following in.

```
akka.management.cluster.health.check.default = "akka.management.HealthCheckRoutes"
akka.management.http.route-providers += ${akka.management.cluster.health.check.default}
```

A user could than disable it by doing:
```
akka.management.cluster.health.check.default = null
```
Or adding another implementation.

But then we will need to filter out `null` and "" in akka-management. That's easy to do.

Of course, that's Lagom/Play needs, but if can find a built-in way to diable routes that will be useful. I personally think it make sense to add it by default because it offers a complete solution. But defaults can't be a lock-in. We should only have it if we have the means to remove it.

**Comment by dwijnand** [comment]
(Option 3: https://twitter.com/dwijnand/status/1096684643871608832 but @viktorklang does not approve 😄)

**Comment by ignasi35** [comment]
IIUC, these are the current defaults:

 * `health-check-routes` - disabled
    * `cluster-health-checks` - enabled (when `health-check-routes` available), though can be [replaced/disabled](https://github.com/akka/akka-management/pull/475)
 * [`cluster-management-routes`](https://github.com/akka/akka-management/blob/b02a0fb1656f060cca8ce775d8a5a8ba06e80d22/cluster-http/src/main/resources/reference.conf#L11)
    * by default `readOnly` but users may opt in to `readWrite`. Can't disable.
 * `http-cluster-bootstrap` - enabled (and required)

Only `health-check-routes` is not enabled by default. 

I am not familiar with `cluster-management-routes` but users might want to disable even the readOnly routes.

`http-cluster-bootstrap` are required so this discussion doesn't apply.

-------------------------------------------------

I think the small differences in the defaults cause some confusion.

**Comment by ignasi35** [comment]
To be constructive, I'd use the following:


 * `health-check-routes` - disabled
    * `cluster-health-checks` - enabled (when `health-check-routes` available)
 * [`cluster-management-routes`] - disabled
    * when enabled, default to `readOnly` (users may opt in to `readWrite`)
 * `http-cluster-bootstrap` - enabled (because it's required)

**Comment by renatocaval** [comment]
I think you are naming two things that don't exist. 

Akka Cluster Bootstrap requires Cluster Http Management (in read-only at least).

**Comment by renatocaval** [comment]
@dwijnand 

> (Option 3: [twitter.com/dwijnand/status/1096684643871608832](https://twitter.com/dwijnand/status/1096684643871608832) but @viktorklang does not approve 😄)

It turns out that I also need it. So now we need it more than once.

**Comment by chbatey** [comment]
> Akka Cluster Bootstrap requires Cluster Http Management (in read-only at least).

Only if you want the health check. No dep otherwise

**Comment by jrudolph** [comment]
> Could we instead add the route by default and add another flag that guards whether the health-check is really enabled or not?

Maybe I should make that idea more concrete.

Do we think `akka.management.http.route-providers` is even the user API that users should use? Or is it more of an SPI where libraries can provide routes?

If we see it as an SPI, each module could unconditionally just add there routes and add another user-facing config option that really enables that route.

We could standardize on a name for that or add another option, so that it's `readOnly`, `readWrite`, or `disabled`. The route providers would still have to manage that setting themselves.

**Comment by jrudolph** [comment]
> We could standardize on a name for that or add another option, so that it's `readOnly`, `readWrite`, or `disabled`. The route providers would still have to manage that setting themselves.

I see that's pretty similar to what's already done in #471.
